tests/handlers/test_e2e_keys.py

```diff
@@ -831,3 +836,94 @@ def test_query_devices_remote_sync(self):
                 }
             },
         )
+
+    @parameterized.expand(
+        [
+            # The remote homeserver's response indicates that this user has 0/1/2 devices.
+            ([],),
+            (["device_1"],),
+            (["device_1", "device_2"],),
+        ]
+    )
+    def test_query_all_devices_caches_result(self, device_ids: Iterable[str]):
+        """Test that requests for all of a remote user's devices are cached.
+
+        We do this by asserting that only one call over federation was made, and that
+        the two queries to the local homeserver produce the same response.
+        """
+        local_user_id = "@test:test"
+        remote_user_id = "@test:other"
+        request_body = {"device_keys": {remote_user_id: []}}
+
+        response_devices = [
+            {
+                "device_id": device_id,
+                "keys": {
+                    "algorithms": ["dummy"],
+                    "device_id": device_id,
+                    "keys": {f"dummy:{device_id}": "dummy"},
+                    "signatures": {device_id: {f"dummy:{device_id}": "dummy"}},
+                    "unsigned": {},
+                    "user_id": "@test:other",
+                },
+            }
+            for device_id in device_ids
+        ]
+
+        response_body = {
+            "devices": response_devices,
+            "user_id": remote_user_id,
+            "stream_id": 12345,  # an integer, according to the spec
+        }
+
+        e2e_handler = self.hs.get_e2e_keys_handler()
+
+        # Pretend we're sharing a room with the user we're querying. If not,
+        # `_query_devices_for_destination` will return early.
+        mock_get_rooms = mock.patch.object(
+            self.store,
+            "get_rooms_for_user",
+            new_callable=mock.MagicMock,
+            return_value=make_awaitable(["some_room_id"]),
+        )
```

Comment:
Why the change from using `patch` directly? I'm surprised this needs `new_callable` here, I've hardly ever seen that used. Why do we need to provide it?

Reply:
> Why the change from using patch directly?

I noticed that the file was already importing `mock` in a few places before I came along. I thought this was slightly more consistent.

> I'm surprised this needs new_callable here, I've hardly ever seen that used. Why do we need to provide it?

The previous version without `new_callable` worked fine on 3.7 but failed on 3.10 (and I'm guessing would have failed on 3.8--3.9). I thought I understood why, but since the tests are all failing I clearly didn't. Argh. Let me take another look.

Reply:
From [the docs](https://docs.python.org/3.7/library/unittest.mock.html#unittest.mock.patch):

> *new_callable* allows you to specify a different class, or callable object, that will be called to create the *new* object. By default `MagicMock` is used.

So this sounds to me like we're just passing in the default.

Reply:
That's only the case in 3.7. In 3.8 and up:

> _new_callable_ allows you to specify a different class, or callable object, that will be called to create the _new_ object. By default `AsyncMock` is used for async functions and `MagicMock` for the rest.

Reply:
Yes, exactly: those are the failures in the checks for 0999ed8.

There's a new failure in 13e0aec caused by me overwriting `stream_id` after the rename! Fixed in 6fbf316, hopefully.

Reply:
Yeah, I think we'll be able to clean-up a lot of code once we're only on 3.8! Thanks for the explanation!

Reply:
(I'd started writing out a wall of text to cover the mechanics in detail, but... fortunately you've seen this awaitable-awaitable before!)